tsfc/driver.py

```diff
@@ -359,19 +359,20 @@ def fn(point_set, deriv=None):
             '''Wrapper function for converting UFL `expression` into GEM expression.
             '''
             # Is exterior derivative required?
-            from ufl.operators import Dx, Dn, div, grad, nabla_div, nabla_grad, curl, \
-                                      rot, diff
-            deriv_dict = {'Dx': Dx, 'Dn': Dn, 'div': div, 'grad': grad, 'nabla_div': nabla_div, \
-                          'nabla_grad': nabla_grad, 'curl': curl, 'rot': rot, 'diff': diff}
+            from ufl.operators import Dn, div, grad, nabla_div, nabla_grad, curl, rot
+            deriv_oper = {'Dn': Dn, 'div': div, 'grad': grad, 'nabla_div': nabla_div,
+                          'nabla_grad': nabla_grad, 'curl': curl, 'rot': rot}
             config = kernel_cfg.copy()
             config.update(point_set=point_set)
             if deriv is None:
                 dexpression = expression
-            elif isinstance(deriv, (tuple, list)):
-                # TODO: How to do Dx and diff?
+            elif type(deriv) is int:
+                # Dx requires GEM indexing in FInAT
                 dexpression = expression
+                for _ in range(deriv):
+                    dexpression = grad(dexpression)
             else:
-                dexpression = deriv_dict[deriv](expression)
+                dexpression = deriv_oper[deriv](expression)
```

Comment:
I _think_ that the interface for specifying the derivatives doesn't need to be this complicated. FInAT is going to be requesting derivatives based on information in the functionals, and they will always want (components) of the gradient (or higher derivatives) of the input function. So I think you just need to be able to specify how many grads to apply to each component of the `ufl_shape` of the input expression and separately which components of the resulting expression to select (or does that selection happen on the "other side")?

Reply:
Same as above.